src/rshim.c

```diff
@@ -2157,30 +2203,183 @@ static int rshim_update_locked_mode(rshim_backend_t *bd)
       int rt;
 
       rshim_lock();
-      pthread_mutex_lock(&bd->mutex);
       rt = rshim_access_check(bd); 
-      pthread_mutex_unlock(&bd->mutex);
       rshim_unlock();
       if (rt) {
         RSHIM_INFO("rshim%d attached by another device. Entering Drop Mode\n",
             bd->index);
+        pthread_mutex_lock(&bd->mutex);
         rshim_set_drop_mode(bd, 1);
+        pthread_mutex_unlock(&bd->mutex);
       }
     }
   }
 
   return 0;
 }
 
+static int read_rshim_sp1(rshim_backend_t *bd, uint64_t* sp1)
+{
+  int rc;
+
+  rc = bd->read_rshim(bd, RSHIM_CHANNEL, bd->regs->scratchpad1, sp1, RSHIM_REG_SIZE_8B);
+  if (rc) {
+    RSHIM_ERR("rshim%d failed to read sp1\n", bd->index);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int check_rshim_sp1_magic(rshim_backend_t *bd, uint64_t magic,
+    bool* result)
+{
+  int rc;
+  uint64_t sp1;
+
+  rc = read_rshim_sp1(bd, &sp1);
+  if (rc) {
+    RSHIM_ERR("rshim%d failed to read sp1: %d\n", bd->index, rc);
+    *result = false;
+    return rc;
+  }
+
+  *result = (sp1 == magic);
+
+  return 0;
+}
+
+static int check_rshim_magic_ownership_req(rshim_backend_t *bd, bool* result)
+{
+  return check_rshim_sp1_magic(bd, RSHIM_OWNERSHIP_REQ_MAGIC_NUM, result);
+}
+
+static int check_rshim_magic_ownership_ack(rshim_backend_t *bd, bool* result)
+{
+  return check_rshim_sp1_magic(bd, RSHIM_OWNERSHIP_ACK_MAGIC_NUM, result);
+}
+
+static int write_rshim_sp1_magic(rshim_backend_t *bd, uint64_t magic)
+{
+  int rc;
+
+  rc = bd->write_rshim(bd, RSHIM_CHANNEL, bd->regs->scratchpad1, magic,
+      RSHIM_REG_SIZE_8B);
+  if (rc) {
+    RSHIM_ERR("rshim%d failed to write sp1: %s\n", bd->index, strerror(-rc));
+    return -ENODEV;
+  }
+
+  return 0;
+}
+
+static int write_rshim_magic_ownership_req(rshim_backend_t *bd)
+{
+  RSHIM_INFO("rshim%d writing ownership req magic: 0x%llx\n", bd->index,
+      RSHIM_OWNERSHIP_REQ_MAGIC_NUM);
+
+  return write_rshim_sp1_magic(bd, RSHIM_OWNERSHIP_REQ_MAGIC_NUM);
+}
+
+static int write_rshim_magic_ownership_ack(rshim_backend_t *bd)
+{
+  RSHIM_INFO("rshim%d writing ownership ack magic 0x%llx\n", bd->index,
+      RSHIM_OWNERSHIP_ACK_MAGIC_NUM);
+
+  return write_rshim_sp1_magic(bd, RSHIM_OWNERSHIP_ACK_MAGIC_NUM);
+}
+
+static int handle_ownership_transfer(rshim_backend_t *bd) {
+  int i, rt;
+  bool has_ack;
+  bool has_req;
+
+  if (bd->drop_mode) {
+    if (bd->force_cmd_pending) {
+      RSHIM_INFO("rshim%d executing Force command\n", bd->index);
```

Comment:
This is the place to enable the device if drop_mode==1 && force_cmd_pending==1, in case pcie is unmapped.

if (bd->enable_device && !bd->enable_device(bd, true))
  return fail.
owner ship transfer

Reply:
Will change. BTW, it seems it should be `if (bd->enable_device && bd->enable_device(bd, true))`. Notice there should be no `!`.